### PR TITLE
Explicitly mark parquet for tests in datafusion-common 

### LINKED ISSUE #8250: Use of undeclared crate or module `parquet` when compiling without `--feature=parquet` flag
### Describe the bug

I believe this breaks after https://github.com/apache/arrow-datafusion/pull/7745, but it has been unsolved for while, so I file an issue to track it. Adding `#[cfg(feature = "parquet")]` is a workaround for now, but if want to run single test inside tests we need to add another flag for it.

### To Reproduce

Run tests in `datafusion/common/src/scalar.rs` like this:

```shell
$ cargo test -p datafusion-common
   Compiling datafusion-common v33.0.0 (/Users/alamb/Software/arrow-datafusion2/datafusion/common)
error[E0433]: failed to resolve: use of undeclared crate or module `parquet`
   --> datafusion/common/src/file_options/mod.rs:302:9
    |
302 |     use parquet::{
    |         ^^^^^^^ use of undeclared crate or module `parquet`

error[E0432]: unresolved import `super::parquet_writer`
   --> datafusion/common/src/file_options/mod.rs:316:17
    |
316 |     use super::{parquet_writer::ParquetWriterOptions, StatementOptions};
    |                 ^^^^^^^^^^^^^^ could not find `parquet_writer` in `super`
```

### Expected behavior

I hope we can have a way to just run the test without adding flag like before. Either all the test at once or single test.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Dennis40816**
## Which issue does this PR close?

Closes #8250.

## What changes are included in this PR?

In the tests for `datafusion-common`, mark the test programs that require parquet for compilation with `#[cfg(feature = "parquet")]`. Additionally, in the CI, add a check to verify whether `datafusion-common` can compile without any features enabled (by `cargo check`).

## Are these changes tested?

Running `cargo test --tests --no-default-features -p datafusion-common`.
The result is a pass.

### CODE REVIEW

**Comment by Dennis40816** [comment]
@alamb, I would appreciate it if you could review this PR to ensure it aligns with our expectations.

File: .github/workflows/rust.yml [hunk 1]
```diff
@@ -66,7 +66,8 @@ jobs:
           key: cargo-cache-benchmark-${{ hashFiles('datafusion/**/Cargo.toml', 'benchmarks/Cargo.toml', 'datafusion-cli/Cargo.toml') }}
 
       - name: Check workspace without default features
-        run: cargo check --no-default-features -p datafusion
+        run: cargo check --no-default-features -p datafusion &&
+             cargo check --tests --no-default-features -p datafusion-common
```
Comment (alamb) [inline .github/workflows/rust.yml:70]: Could we please make this a separate name (to make it easier to see in the UI

Like:

```suggestion
        run: cargo check --no-default-features -p datafusion
        
      - name: Check datafusion-common  without default features
        run: cargo check --tests --no-default-features -p datafusion-common
```

**Review by alamb — APPROVED** [review]
Thank you @Dennis40816  -- I have just one small request and I think this PR will be good to go

File: .github/workflows/rust.yml (hunk 1 above)
Comment (Dennis40816) [reply .github/workflows/rust.yml:70]: No problem at all.

**Review by alamb — APPROVED** [review]
Thanks @Dennis40816 !

cc @jayzhan211

**Comment by alamb** [comment]
The CI check seems to be great:

https://github.com/apache/arrow-datafusion/actions/runs/7180238066/job/19571832278?pr=8497

Screenshot 2023-12-12 at 2 56 41 PM

**Comment by alamb** [comment]
Thanks again @Dennis40816

**Comment by jayzhan211** [comment]
Thanks @Dennis40816 !

**Pull request merged**